### ISSUE TITLE
Make file system API work more like the rest of OSAL

### ISSUE THREAD

**Issue by skliper** [nasa/osal#35]
In addition to the error code issue in ticket #34, other aspects of the filesystem API should be fixed to better integrate with the rest of OSAL:
* The filesystem API is not a full abstraction, it is only a light wrapper around the POSIX/C library functions.  Similarly, the "os_fstat_t", "os_dirp_t", and "os_dirent_t" are directly typedef'ed to the C library types and not abstracted in any way.  The effect of this is that any application code written using the OSAL filesystem API is ultimately still dependent on the underlying system libraries and headers, and the particular POSIX/C standard variant it supported, and may not be portable to other OS's.
* Due to the direct use of OS-supplied types and implementation, this implicitly adds a limitation that any OSAL filesystem must also map to a "real" filesystem recognized by the underlying OS.  It is not possible to access a file storage device unless the OS already recognizes its filesystem.
* The API calls are defined differently and look/feel more like C library functions than the rest of OSAL.  Some FS functions return an "int32" which serves as both an object ID and error code.  Other FS function return a pointer and these have no way of returning a more descriptive error other beyond a NULL pointer.  In the rest of OSAL, API functions pass back an object ID separately from the error code via a separate "uint32" parameter.  This eliminates confusion of how to differentiate an error code from a valid object ID.

This can be fixed without breaking API compatibility with existing code:
* The types which are currently directly typedef'ed to the C library types can be properly abstracted.  This will be transparent to application code as long as the names are preserved.
* A new file API can be introduced that works more like the rest of OSAL, and wrappers implemented to mimic the current API so that existing code still works.  To save code memory space, an optional compile time switch (e.g. in osconfig.h) can disable the wrapper functions once application code is updated.

**Comment by skliper** [comment]
_Imported from trac issue 12.  Created by jphickey on 2015-01-13T13:22:57, last modified: 2019-07-26T15:13:27_

**Comment by skliper** [comment]
_Trac comment by jphickey on 2015-03-18 23:13:39:_

Pushed branch "trac-12-filesys_api", commit [changeset:5a95241]

This is ready for review.  The above change introduces a more complete filesystem abstraction as well as fixing the APIs to work more like the rest of OSAL (e.g. int32 success/error response, uint32 handles, other outputs in pointer arguments).

The new APIs can exist in parallel to the existing API, until it is deemed safe and reasonable to clean up the old API definitions.

**Comment by skliper** [comment]
_Trac comment by jphickey on 2015-04-06 15:02:55:_

This changeset has been rebased to the (unapproved) 2015-04-14 baseline

This is due to a dependency on files added/changed as part of trac #40.  Attempting to base this change on the official 2015-02-26 OSAL is potentially possible but will result in merge conflicts when putting these back together again.

The new commit is [changeset:040426b]

**Comment by skliper** [comment]
_Trac comment by jphickey on 2015-08-17 12:11:10:_

== REBASE ==

This has been rebased to the current `development` branch for review at the next CCB.

Updated commit: [changeset:565131e]

It has also been merged into the `ic-ccb-review` branch accordingly.

**Comment by skliper** [comment]
_Trac comment by sduran on 2015-08-25 11:11:40:_

It looks like "filedes" is an index into the FDTable, not a file descriptor itself, so its type has been changed from int32 to uint32, which is ok.  However, there are a bunch of places where the new uint32 filedes is checked if it is less than 0.

**Comment by skliper** [comment]
_Trac comment by glimes on 2015-08-25 11:24:10:_

Receiving file descriptors in an `int32` and having
to convert them to `uint32` before use does not feel
like a very good idea.

1) data conversion is not tied to error checking by
the language. Negative values are happily converted
into predictable very positive values.

2) service code receiving an OSAL FD still has to
check that it is in range. In fact, the service
code has these checks, and as Steve noted above
while I was typing, it still checks for and rejects
negative values. Of the unsigned input.

3) Implicit conversion of data between signed and
unsigned integers is a significant bug source, and
some of us turn on that warning. This change means
all code that uses OS_open or OS_creat has to now
add explicit conversion of the file descriptor. This
will probably be done by irritated engineers to fix
a warning, who may not take the time to plant proper
error checking, triggering *MORE* discussion of this
API change (and probably hostile discussion) during
code reviews.

I recommend we rethink this, and keep our OSAL file
descriptors in, and pass them around as, int32 values.

**Comment by skliper** [comment]
_Trac comment by acudmore on 2015-08-25 12:35:36:_

I agree with the others that the OS file descriptor int32 vs uint32 needs to be considered. 

But I like the idea of removing the direct use of the posix structures for stat and directories. This will be work well for the smallest systems that use FreeRTOS with minimal C library and file system support.

**Comment by skliper** [comment]
_Trac comment by jphickey on 2015-08-25 12:59:20:_

It has been quite some time since I last looked at this change set, and it is almost like reviewing it for the first time again.

Here is my take on the `uint32` vs. `int32` discussion:
* Currently the file API is an "oddball" within OSAL - every other OSAL component uses `uint32` as its descriptor type (tasks, queues, semaphores, modules, etc).
* All the common code / shared layer features e.g. internal state table management is done based on uint32 as the descriptor type.
* I fully agree that switching between int32 and uint32 is potentially problematic and a source of bugs.

However, I believe that the ability to use shared code to implement the state table management, thereby keeping file descriptor behavior just like the other OSAL components is a benefit that outweighs the risk of switching from `int32` to `uint32`.  

In particular, the actual values returned by the shared layers (the ones that issue object descriptor IDs) are specifically crafted to ensure that they are **NOT** negative if stored in an int32.  This was a consideration when making this change, to be sure that there are no sign conversions where not intended.


In re-examining this code change, the main issue here is that `OS_open` and `OS_create` return an int32 that serves as both an error code (if negative) and a descriptor (if positive).  This is also a diversion from the rest of OSAL that returns a dedicated `int32` error code and returns a descriptor via a `uint32` pointer argument.

I had implemented it this way in hopes of preserving API compatibility with existing apps, accepting the risk of an implicit `int32`->`uint32` conversion.  I believe this to be safe (due to the intentionally limited range of descriptors) and the path of least resistance to get everything using the same type of descriptors.  However, this would actually be more correct and more consistent with the rest of OSAL if the prototypes became e.g.

{{{
int32 OS_open (uint32 *filedes, const char *path, int32 access,uint32 mode);
}}}

This change, however, would create more work for someone upgrading to the new version of OSAL.

**Comment by skliper** [comment]
_Trac comment by glimes on 2015-08-25 13:50:15:_

You present me as a user with a choice; would I prefer:

1. Conversion of the signed value from OS_open into the unsigned
   value needed for OS_read (etc), which I may need to make explicit
   using casts to allow my compiler to warn me about *other* places
   in my code where the signed-unsigned conversion is undesirable;

2. A breaking change to the OS_open API to make it consistant by
   splitting the error return from the filedes return.

As a user, I would prefer to accept that file descriptors are
unique in that they are signed, because they are unique in that
they are combined with negative error codes in the return value.

Or to put it another way... personally, the value I place on
making the type of "filedes" consistant with the types of the
other OSAL items is not large enough to overcome the cost that
this imposes on my code.

**Comment by skliper** [comment]
_Trac comment by cdknight on 2017-10-06 13:25:35:_

+1, this impacts the ability for cFE to, for example, check that an app filename points to a file (not a directory, symlink, etc.)

**Comment by skliper** [comment]
_Trac comment by jphickey on 2018-08-20 16:56:05:_

Note: this needs some sort of review/agreement/disposition because it is actually a prerequisite to #28 (despite the ticket number being higher - it was split off for review reasons).

Commit [changeset:1ce6536] is the latest rebase/revalidation of this effort.  Confirmed that it passes all unit tests.

**Comment by skliper** [comment]
_Trac comment by jphickey on 2018-10-12 09:05:22:_

Ticket Cleanup: this was approved at 2018-08-21 CCB and merged into development branch as of 2018-08-22.

**Comment by skliper** [comment]
_Trac comment by jhageman on 2019-07-03 13:31:58:_

Moved closed 4.2.2 milestone tickets to 4.3.0

**Comment by skliper** [comment]
_Trac comment by jhageman on 2019-07-26 15:13:27:_

Milestone renamed
